ykrt/src/thread_intercept.rs

```diff
@@ -1,11 +1,83 @@
 use std::os::raw::{c_int, c_void};
 
+use std::{ffi::CString, ptr::null_mut};
+
+use libc::{dlsym, free, malloc, pthread_attr_t, pthread_create, pthread_t};
+
+const SHADOW_STACK_SIZE: usize = 1000000;
+
+#[derive(Debug)]
+struct ThreadRoutine {
+    pub args: *mut c_void,
+    pub start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+}
+
+extern "C" fn wrap_thread_routine(arg: *mut c_void) -> *mut c_void {
+    let str = CString::new("shadowstack_0").unwrap();
+    let shadowstack_symbol_addr = unsafe {
+        //  Obtain address of a shadowstack_0 symbol
+        dlsym(null_mut(), str.as_ptr() as *const i8)
+    };
+    if shadowstack_symbol_addr.is_null() {
+        panic!("Unable to find shadowstack address")
+    }
+    let stack_addr = unsafe {
+        // Allocate stack
+        malloc(SHADOW_STACK_SIZE)
+    };
+    if stack_addr.is_null() {
+        panic!("Unable allocate stack")
+    }
+    unsafe {
+        // Set shadowstack symbol with new allocated stack
+        *(shadowstack_symbol_addr as *mut *mut c_void) = stack_addr;
+    }
+    let thread_routine = unsafe {
+        // Obtain ThreadRoutine struct
```

Comment:
Why do we have to `Box` this?

Reply:
We're boxing `ThreadRoutine` cause we're passing a reference from Rust to C that will be used asynchronously.

If we don't Box it, we will have Rust stack-allocated reference, which might be dropped by the time it's used in `__wrap_pthread_create` since `__wrap_pthread_create` and `thread_function` run asynchronously.

I added an explanation in the wrapper function docstring 👉 https://github.com/Pavel-Durov/yk-fork/blob/thread-local-shadowstack/ykrt/src/thread_intercept.rs#L48.

Is there another way of doing it without Boxing?

Reply:
Ah, so the `arg` parameter needs to be carefully documented along the lines of "thread_function: *const c_void is a pointer to a heap value produced by Box::into_raw. This function will consume and drop that box."

Reply:
added 👉 165ad32010100033f9c9db7c28a4e377a560707e